Workaround Linux on ARM event-generation deficiencies

jcpittman144 opened the pull request:
On ARM, Linux does not generate EXECVE_EXIT events, nor does it generate
CLONE_EXIT events to the child process.  libsinsp relies upon these
events to maintain complete and up-to-date threadinfo state.

This PR contains changes, isolated to the specific architecture, to
workaround this issue, by using /proc scanning to fill in threadinfo
state after CLONE and EXEC.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

/kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

> /area driver-kmod

> /area driver-ebpf

> /area libscap

/area libsinsp

> /area tests

> /area proposals

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**What this PR does / why we need it**:
Adds logic in libsinsp event parsing, to workaround the fact that Linux on ARM does not generate the all-important EXECVE_EXIT and CLONE_EXIT (to child) events.  The workaround recognizes the fact that the threadinfo state needs updating, and uses /proc scanning to do so.

Without this workaround, the Sysdig agent (and also Falco) fails to recognize the names of processes after EXECVE, and also fails to associate threads with the proper container.

**Which issue(s) this PR fixes**:
SMAGENT-3462

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of libscap`.
-->

```release-note
NONE
```

alblasco commented:
@jcpittman144 Do you have any docker image, so we can double check that the workaround passes our smoke test? 
https://sysdig-jenkins.internal.sysdig.com/view/QA-internal/job/QA-agent/job/smoke-test/

jcpittman144 commented on userspace/libsinsp/parsers.h:
```diff
@@ -92,6 +92,16 @@ class sinsp_parser
 	bool reset(sinsp_evt *evt);
 	inline void store_event(sinsp_evt* evt);
 
+#if LIBSINSP_CPUARCH_FEATURES != 0
```
My thinking is that all of the (current) CPUARCH_FEATURES require both of
- threadinfo->m_cpuarch_flags
- cpuarch_do_event_processing()

So, instead of polluting the code with a long-winded comparsison all of the specific feature flags, I just check for any CPUARCH_FEATURES configured, for these common CPUARCH-supporting items.

If a future CPUARCH_FEATURE is encountered that changes this current commonality, we can revisit.

speedyguy17 replied:
I see both things going on....we're both trying to solve a problem that's relatively limited in scope, but also making things challenging to extend if we need to.

Perhaps I can offer somewhat of a compromise. It seems to me, there is a disconnect with the scope of what we intend that bitfield to do in this change, and the generality of the name "CPUARCH_FEATURES" (along with some of the other names like CPUARCH_FLAGS, and do_cpuarch_event_processing)

So perhaps I can suggest that we just more narrowly scope the names to "stuff we have to do for unreliable thread info" or some such, it both keeps the code simple, and indicates the mechanism is relatively narrowly scoped, hopefully avoiding the potential issue nathan points out, where we might try to use t for arbitrary arch specific functionality, and thus having unintended consequences of running this code

jcpittman144 commented:
> @jcpittman144 Do you have any docker image, so we can double check that the workaround passes our smoke test? https://sysdig-jenkins.internal.sysdig.com/view/QA-internal/job/QA-agent/job/smoke-test/

I just pushed the following image up to the artifactory:
- docker.internal.sysdig.com/jcpittman-aarch64-agent-exec-debug-image:v2

It is a full agent image, not just an agent-slim or agent-kmodule.

I tried to rerun the specific smoketest that was originally reported as failing:
- https://sysdig-jenkins.internal.sysdig.com/job/QA-agent/job/smoke-test/4846/
But that test run reports HTTP ERROR 404 Not Found, probably was removed after some timeout expired.

speedyguy17 commented on userspace/libsinsp/threadinfo.h:
```diff
@@ -49,6 +50,13 @@ typedef struct erase_fd_params
 
+
+#if LIBSINSP_CPUARCH_FEATURES != 0
+#define LIBSINSP_CPUARCH_FLAG_INIT_CLONE_EXIT_PENDING (1 << 0)
```
why do we restrict defining these only if the features are enabled?

jcpittman144 replied:
This is part of my attempt to minimize the impact of these changes, on any architectures (e.g x86) that do not have CPU architecture-specific features.  Also not present if CPUARCH_FEATURES == 0, is the threadinfo m_cpuarch_flags member that holds these flags.

speedyguy17 replied:
the define should be free though, no?

jcpittman144 replied:
OK, I will move these #defines outside of any #ifdef.

speedyguy17 commented on userspace/libsinsp/threadinfo.h:
```diff
@@ -302,6 +310,11 @@ class SINSP_PUBLIC sinsp_threadinfo
 
+#if LIBSINSP_CPUARCH_FEATURES != 0
+	// State for CPU architecture-specific processing
+	uint32_t m_cpuarch_flags;
```
may i suggest an alternative? It seems a thread inherently has a state regardless of if it is running on x86 or arm (or any architecutre). It's just that in the case of x86, that state is simply "complete" instead of pending.

So I'm not sure we should tie this flags to an architecture. We simply set it to "0" in architectures where those events will not be pending, then we can remove the CPU_ARCH tame and guards on much of this.

Maybe that doesn't make sense for some reason? IDK you know the problem space far better than i.

jcpittman144 replied:
I'm trying to minimize, to nearly 0, the impact on these changes for x86, and other architectures that don't have this nonsense behavior.

speedyguy17 replied:
that's a valid enough reason, I think. It scares me a *bit* when we have more of these things, or deciding whether custom behavior applies to which arch, but I don't have a good intuition about which direction things will ultimatley go...so am hesitant to commit one direction or the other.

speedyguy17 commented on userspace/libsinsp/sinsp.cpp:
```diff
@@ -1523,6 +1523,11 @@ bool sinsp::add_thread(const sinsp_threadinfo *ptinfo)
 	return m_thread_manager->add_thread((sinsp_threadinfo*)ptinfo, false);
 }
 
+void sinsp::reinit_thread_from_proc(sinsp_threadinfo* tinfo)
```
thread manager is a public member of sinsp. The client should directly invoke this on the thread manager instead of needing to use sinsp as a passthru.

speedyguy17 replied:
It's less about efficiency than the long term goal is for sinsp to not simply be a passthru to the things you actually want to talk to, like thread manager and container manager. We saw this a ton in the analyzer code, where the difficulty of isolating components was complicated by the fact that they were using the analyzer as a big API dump which just passed through to other things.

I think the point ultimately is that thread manager and container manager are big enough things that we should probably be dealing with them directly. You can already see this in many places:

    if(threads_from_sinsp)
    {
        m_inspector->m_thread_manager->dump_threads_to_file(m_dumper);


so it's not really super consistent either way. And the two options are to use the thread manager as a standalone module, or make every public API of thread manager also a public API of sinsp. I think ultimatley for modularity, we should just invoke it on thread_manager directly, given that we would love to also do the same thing with the functions you mention down the road.

jcpittman144 replied:
OK, I will remove the pass through function, and have parsers.cpp call the thread_manager function directly.

speedyguy17 commented on userspace/libsinsp/parsers.cpp:
```diff
@@ -911,6 +921,205 @@ void sinsp_parser::register_event_callback(sinsp_pd_callback_type etype, sinsp_p
+	uint32_t restore_cpuarch_flags = tinfo->m_cpuarch_flags;
+
+	// Reinitialize thread state from /proc, to pick up new command line
+	m_inspector->reinit_thread_from_proc(tinfo);
```
my biggest concern here is from a performance perspective. Doing a procfs read for every thread is painful. I suppose if we have no other choice to ship arm, we have no other choice...but I have a suspicion this will be the #1 thing we have to figure out moving forward....whether it's building the proc table from a cache to allow only a single scan or allowing a delayed initialization or whatever.

jcpittman144 replied:
Yes, it is a shame.
One moderating factor: If the parent wins the CLONE_EXIT race, AND if the parent is not running in a container,
then the parent will create the child's threadinfo using info from the parent CLONE_EXIT event, and no /proc reads are needed.  But in the other cases -- child wins the race, or parent IS running in a container, then the added cost of the /proc reads will be incurred.
I can't think of another alternative to working around this ARM Linux bug, without reading from /proc; but I'm open to better ideas.

speedyguy17 replied:
I didn't say I necessarily had better ideas right now....only that I think this will become a major headache moving forward.

jcpittman144 commented:
> Perhaps I can offer somewhat of a compromise...
> more narrowly scope the names to "stuff we have to do for unreliable thread info" or some such,
> it both keeps the code simple, and indicates the mechanism is relative narrowly scoped

OK, I'll try to come up with something that addresses these concerns.

speedyguy17 reviewed (approved):
this is okay by me as is, and trust your judgement in how you and nathan might agree on the flag naming/changing whatever, though would still encourage the change to not passthru sinsp that one API.

jcpittman144 commented on userspace/libsinsp/threadinfo.h:
```diff
@@ -302,6 +306,11 @@ class SINSP_PUBLIC sinsp_threadinfo
 
+#if LIBSINSP_CPUARCH_THREAD_EVENT_BUGS != 0
+	// State for processing of CPU architecture-specific thread event bugs
+	uint32_t m_cpuarch_thread_event_bug_flags;
```
I chose a flags word instead of bools, because sinsp_threadinfo already has an m_flags member for similar (non-CPU-specific) attributes.

jcpittman144 commented on userspace/libsinsp/parsers.cpp:
```diff
@@ -911,6 +921,212 @@ void sinsp_parser::register_event_callback(sinsp_pd_callback_type etype, sinsp_p
+	// The thread may have already been fully-initialized, (e.g. by parent
+	// clone exit).  But if not (e.g. parent in container, or inverted clone),
+	// reinitialize thread state from /proc.
+	if(tinfo->m_comm == "<NA>" && tinfo->m_uid == 0xffffffff)
```
I retained the formatting of this same conditional, lifted from another place in the file.
Like you, I'm in favor of a consistent style; however, since there is no consistent style applied (among files or even within this specific file), I try to be consistent with SOMETHING, in this case I chose 'same logic' over proximity.
